Sources/Aperture/Aperture.swift

```diff
@@ -1,228 +1,662 @@
+				}) else {
+					throw ApertureError.targetNotFound(options.targetId!)
+				}
+				
+				let screenFilter = SCContentFilter(display: display, excludingWindows: [])
+				
+				if let cropRect = options.cropRect {
+					streamConfig.sourceRect = cropRect
+				}
+				
+				if #available(macOS 14.0, *) {
+					streamConfig.width = Int(screenFilter.contentRect.width) * options.scaleFactor
+					streamConfig.height = Int(screenFilter.contentRect.height) * options.scaleFactor
+				} else {
+					streamConfig.width = Int(display.frame.width) * options.scaleFactor
+					streamConfig.height = Int(display.frame.height) * options.scaleFactor
+				}
+				
+				filter = screenFilter
+				break
+			case .window:
+				/// We need to cal this before `SCContentFilter` below otherwise an error is thrown: https://forums.developer.apple.com/forums/thread/743615
+				initializeCGS()
+
+				guard let window = content.windows.first(where: { window in
+					String(window.windowID) == options.targetId
+				}) else {
+					throw ApertureError.targetNotFound(options.targetId!)
+				}
+				
+				let windowFilter = SCContentFilter(desktopIndependentWindow: window)
+				
+				if #available(macOS 14.0, *) {
+					streamConfig.width = Int(windowFilter.contentRect.width) * options.scaleFactor
+					streamConfig.height = Int(windowFilter.contentRect.height) * options.scaleFactor
+				} else {
+					streamConfig.width = Int(window.frame.width) * options.scaleFactor
+					streamConfig.height = Int(window.frame.height) * options.scaleFactor
+				}
+				
+				filter = windowFilter
+				break
+			case .externalDevice:
+				filter = nil
+				break
+			case .audioOnly:
+				guard let display = content.displays.first else {
+					throw ApertureError.noDisplaysConnected
+				}
+				
+				let screenFilter = SCContentFilter(display: display, excludingWindows: [])
+				filter = screenFilter
+				break
+			}
+			
+			if let filter = filter {
+				stream = SCStream(filter: filter, configuration: streamConfig, delegate: self)
+			}
+			
+			if let stream = stream {
+				do {
+					try stream.addStreamOutput(self, type: .screen, sampleHandlerQueue: .global())
+					
+					if options.recordSystemAudio {
+						try stream.addStreamOutput(self, type: .audio, sampleHandlerQueue: .global())
+					}
+					
+					if #available(macOS 15, *), options.microphoneDeviceId != nil  {
+						try stream.addStreamOutput(self, type: .microphone, sampleHandlerQueue: .global())
+					}
+				} catch {
+					throw ApertureError.couldNotStartStream(error)
+				}
+			}
+			
+			do {
+				try await initOutput(target: target, options: options, streamConfig: streamConfig)
+			} catch {
+				try? await cleanUp()
+				throw error is ApertureError ? error : ApertureError.couldNotStartStream(error)
+			}
+		}
+		
+		public func stopRecording() async throws {
+			if let error = error {
+				throw error
 			} else {
-				throw Error.couldNotAddMic
+				onFinish?()
+				try? await cleanUp()
```

Comment:
Should it not clean up on error?

Reply:
The only way we have an error here is if an error was thrown after starting, and `recordError` does the cleanup (in case stop is never called)